File: src/operations/compareAndSet.lua
local valuesHashKey = KEYS[1]
local keysSortedSetKey = KEYS[2]
local versionsSortedSetKey = KEYS[3]
local changesStreamKey = KEYS[4]
local revisionKey = KEYS[5]
local key = ARGV[1]
local compareOperator = ARGV[2]
local compareVersion = ARGV[3]
local value = ARGV[4]
local version = ARGV[5]
local streamMaxLen = ARGV[6]

local currentRevision = redis.call('GET', revisionKey) or '0-0'

local currentVersion = redis.call('ZSCORE', versionsSortedSetKey, key) or '0'

local success

if compareOperator == '<' then
    success = tonumber(currentVersion) < tonumber(compareVersion)
elseif compareOperator == '<=' then
    success = tonumber(currentVersion) <= tonumber(compareVersion)
elseif compareOperator == '==' then
    success = tonumber(currentVersion) == tonumber(compareVersion)
elseif compareOperator == '!=' then
    success = tonumber(currentVersion) ~= tonumber(compareVersion)
elseif compareOperator == '>=' then
    success = tonumber(currentVersion) >= tonumber(compareVersion)
elseif compareOperator == '>' then
    success = tonumber(currentVersion) > tonumber(compareVersion)
else
    success = false
end

if not success then return {currentRevision, success} end

redis.call('HSET', valuesHashKey, key, value)
redis.call('ZADD', keysSortedSetKey, '0', key)
redis.call('ZADD', versionsSortedSetKey, version, key)

local revision = redis.call('XADD', changesStreamKey, 'MAXLEN', '~',
                            streamMaxLen, '*', 'prev', currentRevision, 'key',
                            key, 'value', value, 'version', version)

redis.call('SET', revisionKey, revision)

return {revision, success}

File: src/operations/compareAndSet.ts
import IORedis = require('ioredis');
import * as fs from 'fs';
import * as path from 'path';
import {
  VALUES_HASH_SUFFIX,
  VERSIONS_SORTED_SET_SUFFIX,
  getKey,
  CHANGES_STREAM_SUFFIX,
  REVISION_SUFFIX,
  KEYS_SORTED_SET_SUFFIX,
} from '../keys';
import {CompareOperator} from '../compareOperators';

const definition = {
  numberOfKeys: 5,
  lua: fs.readFileSync(path.join(__dirname, 'compareAndSet.lua'), 'utf-8'),
};

declare module 'ioredis' {
  interface Redis {
    lcCompareAndSetBuffer(
      valuesKey: string,
      keysKey: string,
      versionsKey: string,
      changesKey: string,
      revisionKey: string,
      key: string | Buffer,
      compareOperator: string,
      compareVersion: string,
      value: Buffer,
      version: string,
      maxlen: number,
    ): Promise<[Buffer, null | 1]>;
  }

  interface Cluster {
    lcCompareAndSetBuffer(
      valuesKey: string,
      keysKey: string,
      versionsKey: string,
      changesKey: string,
      revisionKey: string,
      key: string | Buffer,
      compareOperator: string,
      compareVersion: string,
      value: Buffer,
      version: string,
      maxlen: number,
    ): Promise<[Buffer, null | 1]>;
  }

  interface Pipeline {
    lcCompareAndSetBuffer(
      valuesKey: string,
      keysKey: string,
      versionsKey: string,
      changesKey: string,
      revisionKey: string,
      key: string | Buffer,
      compareOperator: string,
      compareVersion: string,
      value: Buffer,
      version: string,
      maxlen: number,
      callback?: IORedis.Callback<[Buffer, null | 1]>,
    ): IORedis.Pipeline;
  }
}

export function defineCompareAndSetCommand(redis: IORedis.Redis | IORedis.Cluster) {
  redis.defineCommand('lcCompareAndSet', definition);
}

export function transformCompareAndSetArguments(
  collection: string,
  key: string | Buffer,
  compareOperator: CompareOperator,
  compareVersion: number,
  value: Buffer,
  version: number = Infinity,
  maxlen: number = 1000,
): Parameters<IORedis.Redis['lcCompareAndSetBuffer']> {
  return [
    getKey(collection, VALUES_HASH_SUFFIX),
    getKey(collection, KEYS_SORTED_SET_SUFFIX),
    getKey(collection, VERSIONS_SORTED_SET_SUFFIX),
    getKey(collection, CHANGES_STREAM_SUFFIX),
    getKey(collection, REVISION_SUFFIX),
    key,
    compareOperator,
    compareVersion.toString(),
    value,
    version.toString(),
    maxlen,
  ];
}

export type CompareAndSetResult = {
  revision: string;
  success: boolean;
};

export function transformCompareAndSetReply(
  reply: [Buffer, null | 1],
): CompareAndSetResult {
  const [revision, success] = reply;

  return {
    revision: revision.toString(),
    success: !!success,
  };
}

export function compareAndSet(
  redis: IORedis.Redis,
  collection: string,
  key: string | Buffer,
  compareOperator: CompareOperator,
  compareVersion: number,
  value: Buffer,
  version: number = Infinity,
  maxlen: number = 1000,
): Promise<CompareAndSetResult> {
  return redis
    .lcCompareAndSetBuffer(
      ...transformCompareAndSetArguments(
        collection,
        key,
        compareOperator,
        compareVersion,
        value,
        version,
        maxlen,
      ),
    )
    .then(transformCompareAndSetReply);
}
